exporter/otlpexporter/arrow.go

```diff
@@ -0,0 +1,469 @@
+	ready streamPrioritizer
+
+	// returning is used to pass broken, gracefully-terminated,
+	// and otherwise to the stream controller.
+	returning chan *arrowStream
+
+	// cancel cancels the background context of this
+	// arrowExporter, used for shutdown.
+	cancel context.CancelFunc
+
+	// wg counts one per active goroutine belonging to all strings
+	// of this exporter.
+	wg sync.WaitGroup
+}
+
+// streamPrioritizer is a placeholder for a configurable mechanism
+// that selects the next stream to write.
+type streamPrioritizer struct {
+	// channel will be closed to downgrade to standard OTLP,
+	// otherwise it returns the first-available.
+	channel chan *arrowStream
+}
+
+// writeItem is passed from the sender (a pipeline consumer) to the
+// stream writer, which is not bound by the sender's context.
+type writeItem struct {
+	// records is a ptrace.Traces, plog.Logs, or pmetric.Metrics
+	records interface{}
+	// errCh is used by the stream reader to unblock the sender
+	errCh chan error
+}
+
+// arrowStream is 1:1 with gRPC stream.
+type arrowStream struct {
+	// client uses the exporter's grpc.ClientConn.
+	client arrowpb.EventsService_EventStreamClient
+
+	// toWrite is passes a batch from the sender to the stream writer, which
+	// includes a dedicated channel for the response.
+	toWrite chan writeItem
+
+	// producer is exclusive to the holder of the stream.
+	producer *batchEvent.Producer
+
+	// cancel cancels the stream context.
+	cancel context.CancelFunc
+
+	// lock protects waiters.
+	lock sync.Mutex
+
+	// waiters is the response channel for each active batch.
+	waiters map[string]chan error
+}
+
+// startArrowExporter creates the background context used by all streams and
+// starts a stream controller, which initializes the initial set of streams.
+func (e *exporter) startArrowExporter() *arrowExporter {
+	bgctx, bgcancel := context.WithCancel(context.Background())
+	bgctx = e.enhanceContext(bgctx)
+
+	ae := &arrowExporter{
+		exporter:  e,
+		client:    arrowpb.NewEventsServiceClient(e.clientConn),
+		ready:     e.newPrioritizer(),
+		returning: make(chan *arrowStream, e.config.Arrow.NumStreams),
+		cancel:    bgcancel,
+	}
+	ae.wg.Add(1)
+	go ae.runStreamController(bgctx)
+
+	return ae
+}
+
+// runStreamController starts the initial set of streams, then waits for streams to
+// terminate one at a time and restarts them.  If streams come back with a nil
+// client (meaning that OTLP+Arrow was not supported by the endpoint), it will
+// not be restarted.
+func (ae *arrowExporter) runStreamController(bgctx context.Context) {
+	defer ae.cancel()
+	defer ae.wg.Done()
+
+	running := ae.exporter.config.Arrow.NumStreams
+
+	// Start the initial number of streams
+	for i := 0; i < running; i++ {
+		ae.wg.Add(1)
+		go ae.runArrowStream(bgctx)
+	}
+
+	for {
+		select {
+		case stream := <-ae.returning:
+			if stream.client != nil {
+				// The stream closed or broken.  Restart it.
+				ae.wg.Add(1)
```

Comment:
I put a comment above the field describing the pattern I've used. 

https://github.com/open-telemetry/experimental-arrow-collector/pull/10/files#diff-f844749c4941a21e703fefd28bb726ac18c980cce0eae535c3c262044e78ef37R60

The way I think about it, it's simply Add(1) before starting any Goroutine and defer a `Done()` call in the goroutine so that all goroutines are waited for.